sdk-actors/src/main/java/io/dapr/actors/runtime/ActorManager.java

```diff
@@ -50,6 +57,7 @@
     this.runtimeContext = runtimeContext;
     this.actorMethods = new ActorMethodInfoMap(runtimeContext.getActorTypeInformation().getInterfaces());
     this.activeActors = Collections.synchronizedMap(new HashMap<>());
+    this.actorReentrancyContext = new ActorReentrancyContext();
```

Comment:
The reentrancy context is per method, not?

Still, I don't think this will work because it would need a way to garbage collect. Imagine that the call stack failed, there would need to be a way to understand that the call stack is over and release the id.

I like the idea of providing the context via an argument to the Actor's method is Reentrancy annotation is present. This way, it remains a stateless solution.

The preferred solution would still be to have this context passed through via Mono. I think we should move this to the next release since it will require more investigation.

Reply:
> The reentrancy context is per method, not?

The context is per actor id/type combo.

> I like the idea of providing the context via an argument to the Actor's method is Reentrancy annotation is present. This way, it remains a stateless solution.

I avoided passing it through via a method because we originally wanted to have the passthrough happen without downstream actor's having to touch it. I don't think we can achieve that just by passing it through so we'd have to make that call about experience.

> Still, I don't think this will work because it would need a way to garbage collect. Imagine that the call stack failed, there would need to be a way to understand that the call stack is over and release the id.

This is actually one of the few places where Reactor was able to help us. On the original call into the actor we can setup a terminal event (currently we just have onSuccess) that clears the ID. This comes in both success and error flavors. If the actual Java runtime crashes, the data is cleared anyway, so I think this is fine.

> The preferred solution would still be to have this context passed through via Mono. I think we should move this to the next release since it will require more investigation.

As for passing the context through via Mono's SubscriberContext, I don't think we're actually using Reactor in a way that this is possible. The SubscriberContext as we need it is available in a single call chain. Since our chain involves a method that can call into other methods (thus creating a nested mono), we lose access to the original pipeline and context in the nested monos.

I agree this change has gotten a lot more complex than anticipated and we shouldn't try and rush it into the release on the day of code freeze.